File: app/api/auth/auth_controller.py
from flask import request
from app.services import logger, response
from .auth_service import AuthService
from app.models.User_model import User


class AuthController:

    @staticmethod
    async def login():
        try:
            credentials = AuthController._get_login_credentials()
            user = await AuthService.login(credentials)
            logger.info(f"user loggedin | {user.to_dict()}")

            token = await AuthService.get_login_token(user)
            return response.login_success(token, user.to_dict())

        except ValueError as e:
            return response.bad_request(f"Error: {e}")

        except Exception as e:
            logger.error(f"had error in item controller: {e}")
            return response.server_error()

    @staticmethod
    async def singup():
        try:
            credentials = AuthController._get_signup_credentials()
            user = await AuthService.signup(credentials)
            logger.info(f"new accout | {user.to_dict()}")

            token = await AuthService.get_login_token(user)
            return response.login_success(token, user.to_dict())

        except ValueError as e:
            return response.bad_request(f"Error: {e}")

        except Exception as e:
            logger.error(f"error in auth controller: {e}")
            return response.server_error()

    @staticmethod
    async def logout():
        try:
            return response.logout_success()
        except Exception as e:
            logger.error(f"error in auth controller: {e}")
            return response.server_error()

    @staticmethod
    def _get_login_credentials():
        try:
            body = request.get_json()
            credentials = User.login_credentials(
                body["username"],
                body["password"]
            )
            return credentials
        except:
            raise ValueError("Missing credentials")

    @staticmethod
    def _get_signup_credentials():
        try:
            body = request.get_json()
            credentials = User.signup_credentials(
                body["username"],
                body["password"],
                body["fullname"],
                body["imgUrl"],
            )
            return credentials
        except:
            raise ValueError("Missing credentials")

File: app/api/auth/auth.controller.go
package auth

import (
	"github.com/gDenisLit/item-server-go/cmd/dtos"
	"github.com/gDenisLit/item-server-go/cmd/models"
	"github.com/gDenisLit/item-server-go/cmd/services"
	"github.com/gofiber/fiber/v2"
)

func OnLogin(ctx *fiber.Ctx) error {

	credentials := new(dtos.LoginDTO)
	err := ctx.BodyParser(credentials)
	if err != nil {
		return ctx.Status(fiber.StatusUnprocessableEntity).JSON(fiber.Map{
			"error": "Invalid credentials object",
		})
	}

	services.Log.Info("Login request", credentials)
	user, err := Login(credentials)
	if err != nil {
		return ctx.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"error": "Invalid username or password",
		})
	}
	return sendMiniUser(ctx, user)
}

func OnSignup(ctx *fiber.Ctx) error {

	credentials := new(dtos.SignupDTO)
	err := ctx.BodyParser(credentials)
	if err != nil {
		return ctx.Status(fiber.StatusUnprocessableEntity).JSON(fiber.Map{
			"error": "Invalid credentials object",
		})
	}

	services.Log.Info("Signup request", credentials)
	user, err := Signup(credentials)
	if err != nil {
		return ctx.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"error": err.Error(),
		})
	}
	return sendMiniUser(ctx, user)
}

func sendMiniUser(ctx *fiber.Ctx, user *models.User) error {
	miniUser := getMiniUser(user)
	loginToken, err := GetLoginToken(miniUser)
	if err != nil {
		services.Log.Debug("error in auth controller:", err)
		return ctx.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Internal server error",
		})
	}

	services.Log.Info("User logged in:", miniUser)
	ctx.Cookie(getCookie(&loginToken))
	return ctx.Status(fiber.StatusOK).JSON(miniUser)
}

func OnLogout(ctx *fiber.Ctx) error {
	return nil
}

func getMiniUser(loggedinUser *models.User) *dtos.UserDTO {
	miniUser := dtos.UserDTO{
		ID:       loggedinUser.ID,
		Username: loggedinUser.Username,
		Fullname: loggedinUser.Fullname,
		ImgUrl:   loggedinUser.ImgUrl,
	}
	return &miniUser
}

func getCookie(loginToken *string) *fiber.Cookie {
	cookie := &fiber.Cookie{
		Name:     "loginToken",
		Value:    *loginToken,
		Secure:   true,
		SameSite: "None",
	}
	return cookie
}
